unity-environment/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs

```diff
@@ -0,0 +1,189 @@
+        public Rigidbody rb;
+        public Vector3 startingPos;
+        public Quaternion startingRot;
+        public WalkerGroundContact groundContactScript;
+    }
+
+    /// <summary>
+    /// Create BodyPart object and add it to dictionary.
+    /// </summary>
+    /// <param name="t"></param>
+    public void SetupBodyPart(Transform t)
+    {
+        BodyPart bp = new BodyPart
+        {
+            rb = t.GetComponent<Rigidbody>(),
+            joint = t.GetComponent<ConfigurableJoint>(),
+            startingPos = t.position,
+            startingRot = t.rotation
+        };
+        bodyParts.Add(t, bp);
+        bp.groundContactScript = t.GetComponent<WalkerGroundContact>();
+    }
+
+    public override void InitializeAgent()
+    {
+        SetupBodyPart(hips);
+        SetupBodyPart(chest);
+        SetupBodyPart(spine);
+        SetupBodyPart(head);
+        SetupBodyPart(thighL);
+        SetupBodyPart(shinL);
+        SetupBodyPart(footL);
+        SetupBodyPart(thighR);
+        SetupBodyPart(shinR);
+        SetupBodyPart(footR);
+        SetupBodyPart(armL);
+        SetupBodyPart(forearmL);
+        SetupBodyPart(handL);
+        SetupBodyPart(armR);
+        SetupBodyPart(forearmR);
+        SetupBodyPart(handR);
+
+    }
+
+    public Quaternion GetJointRotation(ConfigurableJoint joint)
+    {
+        return (Quaternion.FromToRotation(joint.axis, joint.connectedBody.transform.rotation.eulerAngles));
+    }
+
+    /// <summary>
+    /// Add relevant information on each body part to observations.
+    /// </summary>
+    public void BodyPartObservation(BodyPart bp)
+    {
+        var rb = bp.rb;
+        AddVectorObs(bp.groundContactScript.touchingGround ? 1 : 0); // Is this bp touching the ground
+        bp.groundContactScript.touchingGround = false;
+
+        AddVectorObs(rb.velocity);
+        AddVectorObs(rb.angularVelocity);
+        Vector3 localPosRelToHips = hips.InverseTransformPoint(rb.position);
+        AddVectorObs(localPosRelToHips);
+
+        if (bp.joint && (bp.rb.transform != handL && bp.rb.transform != handR))
+        {
+            var jointRotation = GetJointRotation(bp.joint);
+            AddVectorObs(jointRotation); // Get the joint rotation
+        }
+    }
+
+    /// <summary>
+    /// Loop over body parts to add them to observation.
+    /// </summary>
+    public override void CollectObservations()
+    {
+        AddVectorObs(goalDirection);
+        foreach (var item in bodyParts)
+        {
+            BodyPartObservation(item.Value);
+        }
+    }
+
+    /// <summary>
+    /// Apply torque according to defined goal angle and force.
+    /// </summary>
+    public void SetNormalizedTargetRotation(BodyPart bp, float x, float y, float z, float strength)
+    {
+        x = (x + 1f) * 0.5f;
+        y = (y + 1f) * 0.5f;
+        z = (z + 1f) * 0.5f;
+
+        var xRot = Mathf.Lerp(bp.joint.lowAngularXLimit.limit, bp.joint.highAngularXLimit.limit, x);
```

Comment:
Comment on these three lines, particularly why the x-rotation is different from y and z.

Reply:
@Hunter-Unity Can you explain this? I was wondering why X is different myself.